### ISSUE TITLE
Try Virtuoso

### ISSUE THREAD

**Issue by benjamingeer** [dhlab-basel/Knora#602]
```
brew install virtuoso
cd /usr/local/Cellar/virtuoso/7.2.4.2/var/lib/virtuoso/db
virtuoso-t -f
```

* Web interface is http://localhost:8890/conductor/
* Default admin username/password is dba/dba
* [You have to grant SPARQL_UDPATE permission to the SPARQL user](http://wikis.openlinksw.com/VirtuosoWikiWeb/SPARULDROPUpdatePermissionError)
* [SPARQL endpoint](http://vos.openlinksw.com/owiki/wiki/VOS/VOSSparqlProtocol) is http://localhost:8890/sparql/ for queries (and updates?)
* [Graph Store Protocol endpoint](http://vos.openlinksw.com/owiki/wiki/VOS/VirtGraphProtocolCURLExamples) is http://localhost:8890/sparql-graph-crud
* Virtuoso uses [digest auth](http://vos.openlinksw.com/owiki/wiki/VOS/VirtSPARQLProtectSQLDigestAuthentication) instead of basic auth

**Comment by benjamingeer** [comment]
I made a start on `wip/virtuoso-support`. So far the main problem is that Virtuoso doesn't support HTTP Basic authentication. It does support HTTP Digest authentication, but Akka HTTP Client doesn't implement HTTP Digest.

Two options:

* Use another HTTP client library, such as [Apache HttpClient](https://hc.apache.org/httpcomponents-client-ga/), to make the connection to Virtuoso.
* Handle the 401 Unauthorized part of the protocol by generating the necessary digest authorization header ourselves (possibly with help from Apache HttpClient) and sending it in a second request.

Useful examples:

* https://gist.github.com/jirihelmich/9653951
* https://gist.github.com/jindrichmynarz/7707643
* https://stackoverflow.com/a/41362679

Also it seems that the URL path for anything requiring authentication is `/sparql-auth`: http://vos.openlinksw.com/owiki/wiki/VOS/VirtTipsAndTricksDigestAuthAJAX

**Comment by benjamingeer** [comment]
https://github.com/dhlab-basel/Knora/tree/wip/virtuoso-support

**Comment by benjamingeer** [comment]
A nice explanation of HTTP Digest authentication (thanks @lrosenth):

https://www.sitepoint.com/understanding-http-digest-access-authentication/

**Comment by subotic** [comment]
Solved authentication:  basically, follow this documentation but instead of `digest` use `basic`: http://vos.openlinksw.com/owiki/wiki/VOS/VirtSPARQLProtectSQLDigestAuthentication

**Comment by lrosenth** [comment]
Wow!! Congratulation!!

Lukas

On 18 Sep 2017, at 16:31, Ivan Subotic <notifications@github.com<mailto:notifications@github.com>> wrote:


Solved authentication: basically, follow this documentation but instead of digest use basic: http://vos.openlinksw.com/owiki/wiki/VOS/VirtSPARQLProtectSQLDigestAuthentication

**Comment by subotic** [comment]
Some notes:

- http://docs.openlinksw.com/virtuoso/rdfsparqlruleintro/
- http://kidehen.blogspot.ch/2015/07/conceptual-data-virtualization-across.html

Sparql:
```
LOAD <http://www.w3.org/2000/01/rdf-schema#>
```

Sparql:
```
INSERT
   { GRAPH <http://www.w3.org/2000/01/rdf-schema#>
      {  rdfs:subClassOf   a   owl:TransitiveProperty   }
   }
```


Virtuoso stored procedure:
```
rdfs_rule_set
   ( 'incunabula' ,
     'http://www.knora.org/ontology/incunabula'
   ) ;
```

Sparql:
```
DEFINE input:inference <incunabula>

prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix knora-base: <http://www.knora.org/ontology/knora-base#>

SELECT ?resource ?prop ?obj ?isFileValue ?objPred ?objObj
WHERE {
    BIND(IRI("http://data.knora.org/c5058f3a") AS ?resource)

    ?resource rdf:type ?resourceClass ;
        knora-base:isDeleted false .

    ?resourceClass rdfs:subClassOf* knora-base:Resource .

    ?resource ?prop ?obj .
}
```

**Comment by benjamingeer** [comment]
Oh wow!

**Comment by lrosenth** [comment]
Wow wow🤗

Von meinem iPhone gesendet

Am 22.09.2017 um 12:37 schrieb Benjamin Geer <notifications@github.com<mailto:notifications@github.com>>:


Oh wow!

**Comment by subotic** [comment]
Testing queries:

The data:
```
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
INSERT DATA {
	GRAPH <http://example.com/nsgraph> {
		<http://example.com/ns/A> <http://www.w3.org/2000/01/rdf-schema#subClassOf> <http://example.com/ns/B> .
		<http://example.com/ns/B> <http://www.w3.org/2000/01/rdf-schema#subClassOf> <http://example.com/ns/C> .
		<http://example.com/ns/C> <http://www.w3.org/2000/01/rdf-schema#subClassOf> <http://example.com/ns/D> .
		<http://example.com/ns/D> <http://www.w3.org/2000/01/rdf-schema#subClassOf> <http://example.com/ns/TOP> .
        <http://example.com/data/1> rdf:type <http://example.com/ns/A> .
	}
} 

```

Query 1:

```
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
SELECT *
WHERE {
    BIND(IRI("http://example.com/data/1") as ?resource)
    ?resource rdf:type/rdfs:subClassOf* <http://example.com/ns/TOP> .
    ?resource ?p ?o .
}
```

Works as expected in Virtuoso.

Query 2:
```
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
SELECT ?resource ?prop ?obj
WHERE {
   BIND(IRI("http://example.com/data/1") as ?resource)
  
   ?resource rdf:type ?resourceClass .
  
   ?resourceClass rdfs:subClassOf* <http://example.com/ns/TOP> .
  
   ?resource ?prop ?obj .
}
```

Works as expected in Virtuoso (18 rows).

```
prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix knora-base: <http://www.knora.org/ontology/knora-base#>

SELECT ?resource ?prop ?obj
WHERE {
    BIND(IRI("http://data.knora.org/c5058f3a") AS ?resource)

    ?resource rdf:type ?resourceClass ;
        knora-base:isDeleted false .

    ?resourceClass rdfs:subClassOf* knora-base:Resource .

    ?resource ?prop ?obj .
}
```

Gives back only one row back in virtuoso and not 18 as expected.

**Comment by subotic** [comment]
Virtuoso is driving me crazy.

**Comment by subotic** [comment]
If property paths are **not** used (`rdfs:subClassOf` instead of `rdfs:subClassOf*`), then the query returns 18 results as it should:

```
prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix knora-base: <http://www.knora.org/ontology/knora-base#>
prefix xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?resource ?prop ?obj
WHERE {
    BIND(IRI("http://data.knora.org/c5058f3a") AS ?resource)

    ?resource rdf:type ?resourceClass ;
        knora-base:isDeleted "false"^^xsd:boolean .

    ?resourceClass rdfs:subClassOf knora-base:Resource .

    ?resource ?prop ?obj .
}
```

So there seems to be a problem with property paths. I don't know if what we are experiencing is related to https://github.com/openlink/virtuoso-opensource/issues/127.

I will make a small example and post a separate issue. But if the linked issue is any indication, I don't expect that they will fix it anytime soon, as this issue is almost 3 years old and still not fixed.

**Comment by musicEnfanthen** [comment]
But property paths are working for the example data?
What is the one row you get back as a result?

**Comment by subotic** [comment]
A short update. It seems that the query does work in virtuoso:

```
prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix knora-base: <http://www.knora.org/ontology/knora-base#>

SELECT ?prop ?obj ?isFileValue ?objPred ?objObj
WHERE {
    BIND(IRI("http://data.knora.org/c5058f3a") AS ?resource)

    ?resource rdf:type ?resourceClass ;
        knora-base:isDeleted false .

    ?resourceClass rdfs:subClassOf* knora-base:Resource .

    {


        ?prop rdfs:subPropertyOf* knora-base:hasFileValue .
        BIND(true AS ?isFileValue)

        ?resource ?prop ?obj .
        ?obj knora-base:isDeleted false .
        ?obj ?objPred ?objObj .
    }
    UNION
    {


        ?resource ?prop ?obj .

        FILTER NOT EXISTS {
            ?prop rdfs:subPropertyOf* knora-base:resourceProperty .
        }
    }
}
``` 

the same as this query in graphdb:

```
prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix knora-base: <http://www.knora.org/ontology/knora-base#>

SELECT ?prop ?obj ?isFileValue ?objPred ?objObj
WHERE {
    BIND(IRI("http://data.knora.org/c5058f3a") AS ?resource)

    ?resource rdf:type knora-base:Resource ;
        knora-base:isDeleted false .

    {


        ?resource knora-base:hasFileValue ?obj .

        GRAPH <http://www.ontotext.com/explicit> {
            ?resource ?prop ?obj .
        }

        ?obj knora-base:isDeleted false .

        BIND(true AS ?isFileValue)

        GRAPH <http://www.ontotext.com/explicit> {
            ?obj ?objPred ?objObj .
        }
    }
    UNION
    {


        GRAPH <http://www.ontotext.com/explicit> {
            ?resource ?prop ?obj .
        }

        FILTER NOT EXISTS {
            ?resource knora-base:resourceProperty ?obj .
        }
    }
}
```

With two caveats:

1. I needed to restart virtuoso after running the test (loading the data), as only after the restart did the query run correctly. So now I need to figure out how to trigger whatever is triggered on startup after I load the data.

2. Virtuoso returns every triple twice, so some tweaking of the query is necessary. But it returns the same triple as the grapdhdb query does.

**Comment by benjamingeer** [comment]
You are getting close now!

> Virtuoso returns every triple twice

Maybe SELECT DISTINCT would be enough.

**Comment by subotic** [comment]
yes, `DISTINCT` worked. thanks :-)

**Comment by subotic** [comment]
I have signed up for a Virtuoso trial license.

The behavior up until now is the same for the 7.2 opensource and commercial versions.

There is an 8.0 commercial version, but is even worse, as the following does not work:

```
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX knora-base: <http://www.knora.org/ontology/knora-base#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?valueClass ?baseClass


WHERE {
    ?valueClass rdfs:subClassOf* knora-base:Value .

    OPTIONAL {
        ?valueClass rdfs:subClassOf* ?baseClass .
        FILTER (!isBlank(?baseClass))
    }
}
```

**Comment by subotic** [comment]
Another strange behavior (bug or feature). The optional clause in this query removes results.

```
prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix knora-base: <http://www.knora.org/ontology/knora-base#>

SELECT DISTINCT ?prop ?isLinkProp ?isLinkValueProp ?obj ?objPred ?objObj ?predStandoff ?objStandoff
WHERE {
    BIND(IRI("http://data.knora.org/8a0b1e75") as ?resource)

    ?resource rdf:type/rdfs:subClassOf* knora-base:Resource .
    ?resource knora-base:isDeleted false .

    {
        ?prop rdfs:subPropertyOf* knora-base:hasValue .

        FILTER NOT EXISTS {
            ?prop rdfs:subPropertyOf* knora-base:hasFileValue .
        }

        OPTIONAL {
            ?prop rdfs:subPropertyOf* knora-base:hasLinkToValue .
            BIND(true as ?isLinkValueProp)
        }
        
        {
            ?resource ?prop ?obj .
            ?obj knora-base:isDeleted false .
	    ?obj ?objPred ?objObj .

            FILTER(!(?objPred = knora-base:attachedToProject || ?objPred = knora-base:valueHasStandoff))
        }
    }   
}
```

With `optional` virtuoso returns 8 rows, and without 63.

GraphDB returns 46 in both cases.

**Comment by lrosenth** [comment]
Dear Ivan, dear all,

just a question: does Jena/Fuseki act line GraphDb for these queries, or like Virtuoso? This would be interesting to know...

In addition we should get a trial licence of stardog to compare the behavior and to have more knowledge.

@ivan: If You need to buy a developers license for Virtuoso in order to get support from "openLink" (?) – please do!!! We need some solution and all necessary steps should be undertaken. So please let me know if I can help anything!!

Thanks a lot and best Regards,
	Lukas

**Comment by subotic** [comment]
@lrosenth As far as I see, the Virtuoso open source version and the commercial version behave the same. I have a trial license for Virtuoso. I will also get one for Stardog, and try it out. Also, this query runs fine on Fuseki and GraphDB. 

I'm afraid, that the `optional` behavior is a bug, as it goes against the SPARQL spec. Personally, I find Virtuoso very frustrating but also disconcerting because we are talking here about basic SPARQL things that do not work. Not to mention Virtuoso's stability and the tendency for `segmentation faults` for queries that run just fine on GraphDB and Fuseki.

I will keep you posted.

**Comment by lrosenth** [comment]
@subotic yes, I see. If Virtuoso is not up to the task we have to find another solution... Let‘s talk together about the options we have when You have time...

Lukas

Sent from my iPad

On 27 Sep 2017, at 10:23, Ivan Subotic <notifications@github.com<mailto:notifications@github.com>> wrote:


@lrosenth<https://github.com/lrosenth> As far as I see, the Virtuoso open source version and the commercial version behave the same. I have a trial license for Virtuoso. I will also get one for Stardog, and try it out. Also, this query runs fine on Fuseki and GraphDB.

I'm afraid, that the optional behavior is a bug, as it goes against the SPARQL spec. Personally, I find Virtuoso very frustrating but also disconcerting because we are talking here about basic SPARQL things that do not work. Not to mention Virtuoso's stability and the tendency for segmentation faults for queries that run just fine on GraphDB and Fuseki.

I will keep you posted.

**Comment by subotic** [comment]
As discussed, Virtuoso support will be postponed for now. The discussed alternatives are Stardog and AlegroGraph, where both have a community/free version.
